src/admin_ws.rs

```diff
@@ -0,0 +1,187 @@
+use std::{
+    future::Future,
+    sync::{Arc, Mutex},
+};
+
+use holochain_client::{AdminWebsocket, ConductorApiError, ConductorApiResult};
+use tokio::time::{sleep, Duration};
+
+use crate::{HcHttpGatewayError, HcHttpGatewayResult};
+
+const ADMIN_WS_CONNECTION_MAX_RETRIES: usize = 1;
+const ADMIN_WS_CONNECTION_RETRY_DELAY_MS: u64 = 1000;
+
+/// A wrapper around AdminWebsocket that automatically handles reconnection
+/// when the connection is lost due to network issues or other failures.
+#[derive(Clone)]
+pub struct ReconnectingAdminWebsocket {
+    /// The WebSocket URL to connect to
+    url: String,
+    /// The handle to the AdminWebsocket connection
+    connection_handle: Arc<Mutex<Option<AdminWebsocket>>>,
+    /// Current retry attempt counter
+    current_retries: usize,
+}
+
+impl ReconnectingAdminWebsocket {
+    /// Creates a new ReconnectingAdminWebsocket with the specified parameters
+    pub fn new(url: &str) -> Self {
+        ReconnectingAdminWebsocket {
+            url: url.to_string(),
+            connection_handle: Arc::new(Mutex::new(None)),
+            current_retries: 0,
+        }
+    }
+
+    /// Checks if there is an active connection
+    fn is_connected(&self) -> HcHttpGatewayResult<bool> {
+        let connection = self.connection_handle.lock().map_err(|e| {
+            HcHttpGatewayError::InternalError(format!(
+                "Mutex was poisoned during is_connected check: {}",
+                e
+            ))
+        })?;
+
+        Ok(connection.is_some())
+    }
+
+    /// Gets reconnection retry count
+    pub fn get_reconnection_retries(&self) -> usize {
+        self.current_retries
+    }
+
+    /// Ensures that a connection is established before proceeding
+    async fn ensure_connected(&mut self) -> HcHttpGatewayResult<()> {
+        if self.is_connected()? {
+            return Ok(());
+        }
+
+        self.connect().await
+    }
+
+    /// Attempts to connect to the AdminWebsocket
+    ///
+    /// This will make multiple attempts according to the `max_retries` and
+    /// `retry_delay_ms` settings, with exponential backoff between retries.
+    pub async fn connect(&mut self) -> HcHttpGatewayResult<()> {
+        self.current_retries = 0;
+
+        while self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+            match AdminWebsocket::connect(&self.url).await {
```

Comment:
This function is re-entrant and I don't think it should be. The point of this type is to maintain a single admin connection but we're not protecting this operation by locking. 

We only take out a lock if the connection attempt succeeds which means we might connect multiple clients and then assign one and close the others.